nixos/modules/services/networking/crproxy.nix

```
@@ -0,0 +1,172 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.crproxy;
+
+  allowImageListFile = pkgs.writeTextFile {
+    name = "crproxy-allow-image-list";
+    text = lib.strings.concatLines cfg.allowImageList;
+  };
+  useAllowImageList = (lib.length cfg.allowImageList) != 0;
+  blockIPListFile = pkgs.writeTextFile {
+    name = "crproxy-block-ip-list";
+    text = lib.strings.concatLines cfg.blockIPList;
+  };
+  useBlockIPList = (lib.length cfg.blockIPList) != 0;
+  inherit (lib)
+    mkIf
+    mkEnableOption
+    mkOption
+    types
+    concatStringsSep
+    concatLists
+    optionals
+    literalExpression
+    mkPackageOption
+    ;
+in
+{
+  options.services.crproxy = {
+    enable = mkEnableOption "CRProxy, a generic Docker image proxy";
+
+    package = mkPackageOption pkgs "crproxy" { };
+
+    listenAddress = mkOption {
+      default = ":8080";
+      example = literalExpression ''
+        ":8080"
+      '';
+      type = types.str;
+      description = "Address and port to listen on.";
+    };
+
+    behindProxy = mkOption {
+      default = false;
+      example = literalExpression ''
+        true
+      '';
+      type = types.bool;
+      description = "Behind the reverse proxy such as nginx or caddy, which can receive HTTP headers from fronted nginx or caddy. Enable it when enable nginx or caddy as a reverse proxy server.";
+    };
+
+    userpass = mkOption {
+      default = [ ];
+      example = literalExpression ''
+        [ "user1:password@docker.io" "user2:password@ghcr.io" ]
+      '';
+      type = types.listOf types.str;
+      description = "Host and username and password -u user:pwd@host, which can access the registry that needs login.";
```

```suggestion
      description = "Credentials for registries that require authentication.";
```
Format is obvious from example